ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/ProducerTransactionManager.java

```diff
@@ -0,0 +1,161 @@
+import org.apache.kafka.clients.producer.Producer;
+import org.apache.kafka.clients.producer.ProducerRecord;
+import org.apache.kafka.clients.producer.RecordMetadata;
+import org.apache.kafka.common.KafkaException;
+import org.apache.kafka.common.TopicPartition;
+
+/**
+ * Used to handle transactional writes to the command topic
+ */
+public class ProducerTransactionManager {
+
+  private final TopicPartition commandTopicPartition;
+  private final String commandTopicName;
+
+  private final Consumer<CommandId, Command> commandConsumer;
+  private final Producer<CommandId, Command> commandProducer;
+  private final CommandRunner commandRunner;
+
+  public ProducerTransactionManager(
+      final String commandTopicName,
+      final CommandRunner commandRunner,
+      final Map<String, Object> kafkaConsumerProperties,
+      final Map<String, Object> kafkaProducerProperties
+  ) {
+    this.commandTopicPartition = new TopicPartition(
+            Objects.requireNonNull(commandTopicName, "commandTopicName"),
+            0
+    );
+
+    this.commandConsumer = new KafkaConsumer<>(
+        Objects.requireNonNull(kafkaConsumerProperties, "kafkaConsumerProperties"),
+        InternalTopicJsonSerdeUtil.getJsonDeserializer(CommandId.class, true),
+        InternalTopicJsonSerdeUtil.getJsonDeserializer(Command.class, false)
+    );
+
+    this.commandProducer = new KafkaProducer<>(
+        Objects.requireNonNull(kafkaProducerProperties, "kafkaProducerProperties"),
+        InternalTopicJsonSerdeUtil.getJsonSerializer(true),
+        InternalTopicJsonSerdeUtil.getJsonSerializer(false)
+    );
+    this.commandTopicName = Objects.requireNonNull(commandTopicName, "commandTopicName");
+    this.commandRunner = Objects.requireNonNull(commandRunner, "commandRunner");
+  }
+
+  @VisibleForTesting
+  ProducerTransactionManager(
+      final String commandTopicName,
+      final CommandRunner commandRunner,
+      final Consumer<CommandId, Command> commandConsumer,
+      final Producer<CommandId, Command> commandProducer
+  ) {
+    this.commandTopicPartition = new TopicPartition(
+        Objects.requireNonNull(commandTopicName, "commandTopicName"),
+        0
+    );
+    this.commandConsumer = Objects.requireNonNull(commandConsumer, "commandConsumer");
+    this.commandProducer = Objects.requireNonNull(commandProducer, "commandProducer");
+    this.commandTopicName = Objects.requireNonNull(commandTopicName, "commandTopicName");
+    this.commandRunner = Objects.requireNonNull(commandRunner, "commandRunner");
+  }
+
+
+  /** begins transaction */
+  public void begin() {
+    commandConsumer.assign(Collections.singleton(commandTopicPartition));
+    commandProducer.initTransactions();
+    commandProducer.beginTransaction();
+  }
+
+  public void waitForCommandRunner() {
+    final long endOffset = getEndOffset();
+
+    try {
+      int retries = 0;
+      while (commandRunner.getNumCommandProcessed() < endOffset) {
```

Comment:
Is this assuming that every offset in the command topic corresponds to a command? I thought that with transactions these need not necessarily be in sync since transactions write additional messages into the topic (which take up offsets but aren't commands), though I'm having a surprisingly tough time finding docs about this. Would be good to have someone else confirm or deny.

Reply:
When I was testing this locally, I was able to make several transactions in a row without this method getting stuck. If additional messages were taking up offsets, I'd hit the retry limit and have an exception thrown which didn't happen. So that behavior makes me think it doesn't effect commandConsumer.endOffsets() but someone more familiar with this should chime in. I can't seem to find any documentation about these additional messages either.

Reply:
Every transaction will write at least one control message to each partition in the transaction. This control message (commit or abort message) will have an offset. I'm actually surprised that your local testing worked. 

Anothre risk is if we have aborted transactions: and there is a chance of these if an active produce gets fenced by a newer instance coming on line. In this case, the number of valid (non-aborted commands) will always be divergent from the end offsets. 

The docs on how this works can be found here https://cwiki.apache.org/confluence/display/KAFKA/KIP-98+-+Exactly+Once+Delivery+and+Transactional+Messaging

Reply:
I ended up reusing the functionality we already have in place for commandStore.ensureConsumedPast() which uses offsets. position() will get the next offset to fetch so this will ignore the control messages